plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,476 @@
+
+    private final Repository repository;
+
+    private final StorageFacet storageFacet;
+
+    private final MavenFacet mavenFacet;
+
+    private final MavenPathParser mavenPathParser;
+
+    private final MetadataBuilder metadataBuilder;
+
+    private final MetadataUpdater metadataUpdater;
+
+    private final DocumentBuilderFactory documentBuilderFactory;
+
+    public Worker(final Repository repository,
+                  final boolean update,
+                  final String sql,
+                  final Map<String, Object> sqlParams)
+    {
+      this.sql = sql;
+      this.sqlParams = sqlParams;
+      this.repository = repository;
+      this.storageFacet = repository.facet(StorageFacet.class);
+      this.mavenFacet = repository.facet(MavenFacet.class);
+      this.mavenPathParser = mavenFacet.getMavenPathParser();
+      this.metadataBuilder = new MetadataBuilder();
+      this.metadataUpdater = new MetadataUpdater(update, repository);
+      this.documentBuilderFactory = DocumentBuilderFactory.newInstance();
+    }
+
+    /**
+     * Method rebuilding metadata that performs the group level processing. It uses memory conservative "async" SQL
+     * approach, and calls {@link #rebuildMetadataInner(String, String, Set)} method as results are arriving.
+     */
+    public void rebuildMetadata(final ODatabaseDocumentTx db)
+    {
+      db.command(
+          new OSQLAsynchQuery<ODocument>(
+              sql,
+              new OCommandResultListener()
+              {
+                String currentGroupId = null;
+
+                @Override
+                public boolean result(Object iRecord) {
+                  final ODocument doc = (ODocument) iRecord;
+                  final String groupId = doc.field("groupId", OType.STRING);
+                  final String artifactId = doc.field("artifactId", OType.STRING);
+                  final Set<String> baseVersions = doc.field("baseVersions", OType.EMBEDDEDSET);
+
+                  final boolean groupChange = !Objects.equals(currentGroupId, groupId);
+                  if (groupChange) {
+                    if (currentGroupId != null) {
+                      rebuildMetadataExitGroup(currentGroupId);
+                    }
+                    currentGroupId = groupId;
+                    metadataBuilder.onEnterGroupId(groupId);
+                  }
+                  rebuildMetadataInner(groupId, artifactId, baseVersions);
+                  return true;
+                }
+
+                @Override
+                public void end() {
+                  if (currentGroupId != null) {
+                    rebuildMetadataExitGroup(currentGroupId);
+                  }
+                }
+              }
+          )
+      ).execute(sqlParams);
+    }
+
+    /**
+     * Process exits from group level, executed in isolation.
+     */
+    private void rebuildMetadataExitGroup(final String currentGroupId) {
+      isolate(new Runnable()
```

Comment:
Interesting experiment with working outcome: no need for multiple DBs, nested TXes, nor thread-local monkeying: https://github.com/sonatype/nexus-oss/pull/1327